restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.web.docsapi.service;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import hu.akarnokd.rxjava2.operators.ExpandStrategy;
+import hu.akarnokd.rxjava2.operators.FlowableTransformers;
+import io.reactivex.BackpressureStrategy;
+import io.reactivex.Flowable;
+import io.reactivex.FlowableOnSubscribe;
+import io.reactivex.Single;
+import io.stargate.db.ImmutableParameters;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.datastore.Row;
+import io.stargate.db.query.builder.AbstractBound;
+import io.stargate.db.query.builder.BuiltSelect;
+import io.stargate.db.schema.Column;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+/** Executes pre-built document queries, groups document rows and manages document pagination. */
+public class QueryExecutor {
+  private final Accumulator TERM = new Accumulator();
+
+  private final DataStore dataStore;
+
+  public QueryExecutor(DataStore dataStore) {
+    this.dataStore = dataStore;
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    return queryDocs(1, query, pageSize, pagingState);
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      int identityDepth, AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    BuiltSelect select = (BuiltSelect) query.source().query();
+    if (identityDepth < 1 || identityDepth > select.table().primaryKeyColumns().size()) {
+      throw new IllegalArgumentException("Invalid document identity depth: " + identityDepth);
+    }
+
+    List<Column> idColumns = select.table().primaryKeyColumns().subList(0, identityDepth);
+
+    return execute(query, pageSize, pagingState)
+        .concatMap(rs -> Flowable.fromIterable(seeds(rs, idColumns)), 1)
+        .concatWith(Single.just(TERM))
+        .scan(Accumulator::combine)
```

Comment:
Ok this seems so suspicious to me.. I can confirm that at least this does not work for the first element as it would accumulate some first row twice.. Can we go through this together?

Reply:
OK I get it now, the rows are always in order for the same document.. Is this guaranteed? Isn't there a way simpler way to do this, with some kind of to multimap operator?

Reply:
OK forget it, it's not so easy, but I understood now.

Reply:
RxJavaExceptions might have some tools for this, but the predicate will have to be stateful anyway... Might still be cleaner overall. I'll look more into this.